<a href="https://colab.research.google.com/github/Esaipriya/AISECT-ML/blob/main/HEART_DISEASE_PRED.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import sklearn

In [2]:
dataset=pd.read_csv('/content/heart.csv')
dataset

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
0     40   M           ATA        140          289          0     Normal   
1     49   F           NAP        160          180          0     Normal   
2     37   M           ATA        130          283          0         ST   
3     48   F           ASY        138          214          0     Normal   
4     54   M           NAP        150          195          0     Normal   
..   ...  ..           ...        ...          ...        ...        ...   
913   45   M            TA        110          264          0     Normal   
914   68   M           ASY        144          193          1     Normal   
915   57   M           ASY        130          131          0     Normal   
916   57   F           ATA        130          236          0        LVH   
917   38   M           NAP        138          175          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0      172              N      0.0       Up             0  
1      156              N      1.0     Flat             1  
2       98              N      0.0       Up             0  
3      108              Y      1.5     Flat             1  
4      122              N      0.0       Up             0  
..     ...            ...      ...      ...           ...  
913    132              N      1.2     Flat             1  
914    141              N      3.4     Flat             1  
915    115              Y      1.2     Flat             1  
916    174              N      0.0     Flat             1  
917    173              N      0.0       Up             0  

[918 rows x 12 columns]

In [3]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 918 entries, 0 to 917
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             918 non-null    int64  
 1   Sex             918 non-null    object 
 2   ChestPainType   918 non-null    object 
 3   RestingBP       918 non-null    int64  
 4   Cholesterol     918 non-null    int64  
 5   FastingBS       918 non-null    int64  
 6   RestingECG      918 non-null    object 
 7   MaxHR           918 non-null    int64  
 8   ExerciseAngina  918 non-null    object 
 9   Oldpeak         918 non-null    float64
 10  ST_Slope        918 non-null    object 
 11  HeartDisease    918 non-null    int64  
dtypes: float64(1), int64(6), object(5)
memory usage: 86.2+ KB


# **Heart Disease Dataset Column Details**

1.Age -Patient’s age in years.

2.Sex -M = Male, F = Female.

3.ChestPainType -TA = Typical Angina,
ATA = Atypical Angina, NAP = Non-Anginal Pain,ASY = Asymptomatic

4.RestingBP -Normal: 120 mmHg,
High: ≥140 mmHg

5.Cholesterol	-(in mg/dl),
Normal: <200,
Borderline: 200–239,

High: ≥240.

6.FastingBS- 0 = <120 mg/dl, 1 = ≥120 mg/dl

7.RestingECG- Normal = Normal ECG,ST = ST-T wave abnormality,LVH = Left Ventricular Hypertrophy

8.MaxHR -Highest heart rate, Typical Range: 60–202
9.ExerciseAngina- Y = Yes, N = No

10.Oldpeak-Measures change in ST segment on ECG after exercise. A higher value indicates heart stress

11.ST_Slope	= Up = normal response,
• Flat = moderate risk,
• Down = high risk of heart disease.

12.HeartDisease =  0 = No heart disease,
1 = Heart disease present

In [4]:
dataset.head()

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA        140          289          0     Normal    172   
1   49   F           NAP        160          180          0     Normal    156   
2   37   M           ATA        130          283          0         ST     98   
3   48   F           ASY        138          214          0     Normal    108   
4   54   M           NAP        150          195          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0

In [5]:
dataset.isnull().sum()

Age               0
Sex               0
ChestPainType     0
RestingBP         0
Cholesterol       0
FastingBS         0
RestingECG        0
MaxHR             0
ExerciseAngina    0
Oldpeak           0
ST_Slope          0
HeartDisease      0
dtype: int64

In [6]:
dataset['Sex'].value_counts()

Sex
M    725
F    193
Name: count, dtype: int64

In [7]:
from sklearn.preprocessing import OneHotEncoder
# Apply one-hot encoding to categorical columns
dataset = pd.get_dummies(dataset, columns=["Sex"], drop_first=True,dtype=int)

# Display the first few rows of the encoded dataset
display(dataset.head())

Age ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40           ATA        140          289          0     Normal    172   
1   49           NAP        160          180          0     Normal    156   
2   37           ATA        130          283          0         ST     98   
3   48           ASY        138          214          0     Normal    108   
4   54           NAP        150          195          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  Sex_M  
0              N      0.0       Up             0      1  
1              N      1.0     Flat             1      0  
2              N      0.0       Up             0      1  
3              Y      1.5     Flat             1      0  
4              N      0.0       Up             0      1

In [8]:
dataset['ChestPainType'].value_counts()

ChestPainType
ASY    496
NAP    203
ATA    173
TA      46
Name: count, dtype: int64

In [9]:
from sklearn.preprocessing import LabelEncoder

# Initialize LabelEncoder
label_encoder = LabelEncoder()

# Apply label encoding to the 'ChestPainType' column and convert to int
dataset['ChestPainType_encoded'] = label_encoder.fit_transform(dataset['ChestPainType']).astype(int)

# Display the updated dataframe with the new encoded column
display(dataset.head())

Age ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40           ATA        140          289          0     Normal    172   
1   49           NAP        160          180          0     Normal    156   
2   37           ATA        130          283          0         ST     98   
3   48           ASY        138          214          0     Normal    108   
4   54           NAP        150          195          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  Sex_M  ChestPainType_encoded  
0              N      0.0       Up             0      1                      1  
1              N      1.0     Flat             1      0                      2  
2              N      0.0       Up             0      1                      1  
3              Y      1.5     Flat             1      0                      0  
4              N      0.0       Up             0      1                      2

In [10]:
dataset['RestingECG'].value_counts()

RestingECG
Normal    552
LVH       188
ST        178
Name: count, dtype: int64

In [11]:
# Initialize LabelEncoder
label_encoder = LabelEncoder()

# Apply label encoding to the 'ChestPainType' column and convert to int
dataset['RestingECG_encoded'] = label_encoder.fit_transform(dataset['RestingECG']).astype(int)

# Display the updated dataframe with the new encoded column
display(dataset.head())

Age ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40           ATA        140          289          0     Normal    172   
1   49           NAP        160          180          0     Normal    156   
2   37           ATA        130          283          0         ST     98   
3   48           ASY        138          214          0     Normal    108   
4   54           NAP        150          195          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  Sex_M  \
0              N      0.0       Up             0      1   
1              N      1.0     Flat             1      0   
2              N      0.0       Up             0      1   
3              Y      1.5     Flat             1      0   
4              N      0.0       Up             0      1   

   ChestPainType_encoded  RestingECG_encoded  
0                      1                   1  
1                      2                   1  
2                      1                   2  
3                      0                   1  
4                      2                   1

In [12]:
dataset['ExerciseAngina'].value_counts()

ExerciseAngina
N    547
Y    371
Name: count, dtype: int64

In [13]:
# Apply one-hot encoding to categorical column
dataset = pd.get_dummies(dataset, columns=["ExerciseAngina"], drop_first=True,dtype=int)

# Display the first few rows of the encoded dataset
display(dataset.head())

Age ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40           ATA        140          289          0     Normal    172   
1   49           NAP        160          180          0     Normal    156   
2   37           ATA        130          283          0         ST     98   
3   48           ASY        138          214          0     Normal    108   
4   54           NAP        150          195          0     Normal    122   

   Oldpeak ST_Slope  HeartDisease  Sex_M  ChestPainType_encoded  \
0      0.0       Up             0      1                      1   
1      1.0     Flat             1      0                      2   
2      0.0       Up             0      1                      1   
3      1.5     Flat             1      0                      0   
4      0.0       Up             0      1                      2   

   RestingECG_encoded  ExerciseAngina_Y  
0                   1                 0  
1                   1                 0  
2                   2                 0  
3                   1                 1  
4                   1                 0

In [14]:
dataset.drop(['RestingECG'],axis=1,inplace=True)

In [15]:
dataset['ST_Slope'].value_counts()

ST_Slope
Flat    460
Up      395
Down     63
Name: count, dtype: int64

In [16]:
# Initialize LabelEncoder
label_encoder = LabelEncoder()

# Apply label encoding to the 'ChestPainType' column and convert to int
dataset['ST_Slope_encoded'] = label_encoder.fit_transform(dataset['ST_Slope']).astype(int)

# Display the updated dataframe with the new encoded column
display(dataset.head())

Age ChestPainType  RestingBP  Cholesterol  FastingBS  MaxHR  Oldpeak  \
0   40           ATA        140          289          0    172      0.0   
1   49           NAP        160          180          0    156      1.0   
2   37           ATA        130          283          0     98      0.0   
3   48           ASY        138          214          0    108      1.5   
4   54           NAP        150          195          0    122      0.0   

  ST_Slope  HeartDisease  Sex_M  ChestPainType_encoded  RestingECG_encoded  \
0       Up             0      1                      1                   1   
1     Flat             1      0                      2                   1   
2       Up             0      1                      1                   2   
3     Flat             1      0                      0                   1   
4       Up             0      1                      2                   1   

   ExerciseAngina_Y  ST_Slope_encoded  
0                 0                 2  
1                 0                 1  
2                 0                 2  
3                 1                 1  
4                 0                 2

In [17]:
dataset.drop(['ST_Slope'],axis=1,inplace=True)

The Dataset after done a encode

In [18]:
dataset

Age ChestPainType  RestingBP  Cholesterol  FastingBS  MaxHR  Oldpeak  \
0     40           ATA        140          289          0    172      0.0   
1     49           NAP        160          180          0    156      1.0   
2     37           ATA        130          283          0     98      0.0   
3     48           ASY        138          214          0    108      1.5   
4     54           NAP        150          195          0    122      0.0   
..   ...           ...        ...          ...        ...    ...      ...   
913   45            TA        110          264          0    132      1.2   
914   68           ASY        144          193          1    141      3.4   
915   57           ASY        130          131          0    115      1.2   
916   57           ATA        130          236          0    174      0.0   
917   38           NAP        138          175          0    173      0.0   

     HeartDisease  Sex_M  ChestPainType_encoded  RestingECG_encoded  \
0               0      1                      1                   1   
1               1      0                      2                   1   
2               0      1                      1                   2   
3               1      0                      0                   1   
4               0      1                      2                   1   
..            ...    ...                    ...                 ...   
913             1      1                      3                   1   
914             1      1                      0                   1   
915             1      1                      0                   1   
916             1      0                      1                   0   
917             0      1                      2                   1   

     ExerciseAngina_Y  ST_Slope_encoded  
0                   0                 2  
1                   0                 1  
2                   0                 2  
3                   1                 1  
4                   0                 2  
..                ...               ...  
913                 0                 1  
914                 0                 1  
915                 1                 1  
916                 0                 1  
917                 0                 2  

[918 rows x 13 columns]

In [19]:
from sklearn.model_selection import train_test_split

# Define features (X) and target (y)
X = dataset[['Age','RestingBP','Cholesterol','FastingBS','MaxHR','Oldpeak','ChestPainType_encoded','Sex_M','RestingECG_encoded','ExerciseAngina_Y','ST_Slope_encoded']]
y = dataset['HeartDisease']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (734, 11)
Shape of X_test: (184, 11)
Shape of y_train: (734,)
Shape of y_test: (184,)


In [20]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

In [21]:
# Initialize and train the Logistic Regression model
model = LogisticRegression(max_iter=1000)
model.fit(X_train, y_train)

LogisticRegression(max_iter=1000)

In [22]:
# Make predictions on the test set
y_pred = model.predict(X_test)

In [23]:
# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy:.2f}")

Accuracy: 0.84


In [24]:
# Print classification report
print("Classification Report:")
print(classification_report(y_test, y_pred))

Classification Report:
              precision    recall  f1-score   support

           0       0.77      0.88      0.82        77
           1       0.91      0.81      0.86       107

    accuracy                           0.84       184
   macro avg       0.84      0.85      0.84       184
weighted avg       0.85      0.84      0.84       184



In [25]:
# Make predictions on the test set
y_pred = model.predict(X_test)

In [26]:
import pickle
filename="LogisticRegression.sav"
pickle.dump(model,open(filename,"wb"))

In [27]:
load_model=pickle.load(open("/content/LogisticRegression.sav","rb"))

In [28]:
result=load_model.predict([[37,130,283,0,98,0.0,1,1,2,0,2]])

/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(




result
0=No,
1=Yes



In [29]:
result

array([0])